Abstract layerstore from pull/push distribution code

dmcgowan opened the pull request:
Allows using distribution code without a layer store to support alternate backends (such as for plugins) or wrapping the layer store to support artifact storage.

This is needed to allow plugins to use the distribution pull/push code and reconcile the existing forked push/pull code for plugins. Since plugins will also do artifact storage made appropriate changes to ensure that compression is not handled transparently on upload/download.

Since this code could now be used for plugins, changed from blacklisting plugins on pull to whitelisting specific media types in the pull configuration.

**Note** v1 push code has minimal set of changes since, it will be deprecated soon and will not support any of the new use cases for the distribution code.

ping @tonistiigi @aaronlehmann @stevvooe

aaronlehmann commented on distribution/push.go:
```diff
@@ -59,6 +62,19 @@ type Pusher interface {
 	Push(ctx context.Context) error
 }
 
+type PushLayerProvider interface {
```
This and the other exported types here will need comments to pass lint.

dmcgowan commented on distribution/push_v2.go:
```diff
@@ -425,26 +417,17 @@ func (pd *v2PushDescriptor) uploadUsingSession(
 	layerUpload distribution.BlobWriter,
 ) (distribution.Descriptor, error) {
-	arch, err := pd.layer.TarStream()
+	reader, release, err := pd.layer.Content(ctx, progressOutput, pd.ID(), schema2.MediaTypeLayer)
```
ideally progress output would be done on the returned reader but we are returning a compressed stream but reporting progress on the uncompressed stream. Maybe it would be better to just have a compression flag  and let it return an `io.ReadCloser` and `int64` size for progress.

aaronlehmann replied:
How would it know the size before compressing the payload?

aaronlehmann replied:
Not sure if this is what you were originally suggesting, but I was thinking it might make sense to just have `Content` return the payload in the format it has (i.e. with compression if using a blob store, without compression if using `TarStream`), and indicate to the caller whether compression is applied. The push code would wrap the stream with a compressor if necessary. I think this would simplify the progress monitoring. Currently `storeLayer`'s `Content` method creates a progress reader with the hardcoded message `Pushing`, which doesn't feel right.

dmcgowan replied:
I agree, though this does effect manifest building vs using an existing manifest. I think we can treat that separately and not have it effect this interface. 

This has already got a big bloated on the return value. Might be able to extract this
`Content(ctx context.Context, progressOutput progress.Output, progressKey string, mediaType string) (io.ReadCloser, func(), error)`

If we go that route we can probably just split it out into
```
Open() (io.ReadCloser, error)
Size() int64
MediaType() string
```

Then have the release function either integrated into the closer or register a function to be called on the existing `Release` method.

aaronlehmann replied:
So much better.

aaronlehmann commented:
```
distribution/push_v2_test.go:392: cannot use layer.EmptyLayer (type *layer.emptyLayer) as type PushLayer in field value:
```

dmcgowan committed:
Update distribution vendor for manifest builder change

Manifest builder allows setting the configuration type
for the manifest being build. Additionally the default
type has been renamed to reflect it is an image type.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan commented:
Interface updated and tests fixed, PTAL

A distribution vendor change was needed to support setting config media type.

aaronlehmann commented on distribution/config.go:
```diff
@@ -0,0 +1,233 @@
+
+func (l *storeLayer) MediaType() string {
+	// layer store always returns uncompressed tars
+	return "application/vnd.docker.image.rootfs.diff.tar"
```
Let's define this as a constant. `MediaTypeUncompressedLayer`?

aaronlehmann replied:
Sure, might as well. I don't think backward compatibility for allowing this in manifests is difficult, actually, if we add support in 1.13. Previous versions didn't check the media type, and I believe they autodetect compression. We actually had a bug on master for awhile in the 1.10 timeframe where blobs were not getting compressed at all, and it took a long time to notice because engines tolerate this fine.

dmcgowan replied:
https://github.com/docker/distribution/pull/2110

aaronlehmann commented on distribution/pull_v2.go:
```diff
@@ -355,8 +354,19 @@ func (p *v2Puller) pullV2Tag(ctx context.Context, ref reference.Named) (tagUpdat
+			if configClass == "" {
+				configClass = "unknown"
+			}
+			return false, fmt.Errorf("target is %s", configClass)
```
You'll need to update `cli/command/image/pull.go` to grep for this new error message.

dmcgowan replied:
ahh thanks, I thought it was the same message but `target is a plugin` vs `target is plugin`

aaronlehmann commented on distribution/pull_v2.go:
```diff
@@ -551,84 +566,87 @@ func (p *v2Puller) pullSchema2(ctx context.Context, ref reference.Named, mfst *s
-			return "", "", err
-		default:
-			cancel()
+	if p.config.DownloadManager != nil {
```
Why would `p.config.DownloadManager` ever be nil? Is this configuration useful?

tonistiigi replied:
Plugins pull config only to get the privileges.

aaronlehmann commented on distribution/push_v2.go:
```diff
@@ -425,26 +422,34 @@ func (pd *v2PushDescriptor) uploadUsingSession(
-	defer func() {
+	var (
+		reader io.ReadCloser
+	)
```
nit: `var reader io.ReadCloser`

aaronlehmann commented on distribution/config.go:
```diff
@@ -0,0 +1,233 @@
+}
+
+func (l *storeLayer) Size() int64 {
+	size, _ := l.Layer.DiffSize()
```
I'm a bit conflicted, but I think it's a good idea to pass the error through. At least one place uses this right now (`getMaxMountAndExistenceCheckAttempts`).

dmcgowan committed:
Abstract distribution interfaces from image specific types

Move configurations into a single file.
Abstract download manager in pull config.
Add supports for schema2 only and schema2 type checking.
Add interface for providing push layers.
Abstract image store to generically handle configurations.

Signed-off-by: Derek McGowan <derek@mcgstyle.net>

dmcgowan commented:
CI is passing now, the windowsRS1 is failing because of https://github.com/docker/docker/pull/29494.

Can I get a +1 or -1 on design or move to code review :)

aaronlehmann commented:
Moved to code review.

tonistiigi commented:
LGTM

aaronlehmann commented:
LGTM

thaJeztah commented:
Windows is failing on the DCO;

```
21:58:23 These commits do not have a proper 'Signed-off-by:' marker:
21:58:23  - c609a38639f9a0d20c8922b762336a19c8cb1c2e
21:58:23  - aa010824d6336b030587fe53b51b4b1d0b58032d
21:58:23 Please amend each commit to include a properly formatted DCO marker.
21:58:23 Visit the following URL for information about the Docker DCO:
21:58:23     https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-wor 
21:58:23    k
```

That should be fixed by https://github.com/docker/docker/pull/29494. I think CI does a rebase, so I'll restart windows to see if it fixes it

dmcgowan commented:
Fixed, ci is green now

stevvooe commented on distribution/config.go:
```diff
@@ -0,0 +1,233 @@
+// PushLayer is a pushable layer with metadata about the layer
+// and access to the content of the layer.
+type PushLayer interface {
+	ChainID() layer.ChainID
```
How can a layer know its own `ChainID`? This is really a property of the application order.

The "unpacked" layer should be keyed by the `ChainID`, but the actual layer id may end up being different.

dmcgowan replied:
For a layer push, the layer should know its digest and its parent, which is all that is needed for chain id. I agree though we shouldn't have chain id in this interface, since this was only the first step in refactoring I choose to kept the interface as close to what is currently. But ideally this should either just go away or become a string layer id.
